hypervisor/arch/x86/page.c

```diff
@@ -0,0 +1,172 @@
+		static struct page prefix ## lvl ## _pages[LVL ## _PAGE_NUM(address_space)]
+
+DEFINE_PGTABLE_PAGE(hv_, pml4, PML4, CONFIG_HV_ADDRESS_SPACE);
+DEFINE_PGTABLE_PAGE(hv_, pdpt, PDPT, CONFIG_HV_ADDRESS_SPACE);
+DEFINE_PGTABLE_PAGE(hv_, pd, PD, CONFIG_HV_ADDRESS_SPACE);
+static struct page hv_pt_pages[(E820_MAX_ENTRIES + 1UL) << 1U];
+
+static int64_t hv_pt_page_idx = -1UL;
+
+struct pgtable_pages_info hv_pgtable_pages_info = {
+	.nworld_pml4_base = hv_pml4_pages,
+	.nworld_pdpt_base = hv_pdpt_pages,
+	.nworld_pd_base = hv_pd_pages,
+	.nworld_pt_base = hv_pt_pages,
+};
+
+static inline uint64_t hv_get_default_attr(void)
+{
+	return PAGE_TABLE;
+}
+
+static inline uint64_t hv_pgentry_present(uint64_t pte)
+{
+	return pte & PAGE_PRESENT;
+}
+
+static inline struct page *hv_get_pml4_page(const struct pgtable_pages_info *info, __unused uint64_t gpa)
+{
+	struct page *page = info->nworld_pml4_base;
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static inline struct page *hv_get_pdpt_page(const struct pgtable_pages_info *info, uint64_t gpa)
+{
+	struct page *page = info->nworld_pdpt_base + (gpa >> PML4E_SHIFT);
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static inline struct page *hv_get_pd_page(const struct pgtable_pages_info *info, uint64_t gpa)
+{
+	struct page *page = info->nworld_pd_base + (gpa >> PDPTE_SHIFT);
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+static inline struct page *hv_get_pt_page(const struct pgtable_pages_info *info, __unused uint64_t gpa)
+{
+	struct page *page = info->nworld_pt_base + atomic_inc64_return(&hv_pt_page_idx);
+	(void)memset(page, 0U, PAGE_SIZE);
+	return page;
+}
+
+const struct memory_ops hv_mem_ops = {
+	.get_default_attr = hv_get_default_attr,
+	.pgentry_present = hv_pgentry_present,
+	.get_pml4_page = hv_get_pml4_page,
+	.get_pdpt_page = hv_get_pdpt_page,
+	.get_pd_page = hv_get_pd_page,
+	.get_pt_page = hv_get_pt_page,
+};
+
+DEFINE_PGTABLE_PAGE(vm0_, pml4, PML4, CONFIG_SOS_ADDRESS_SPACE);
+DEFINE_PGTABLE_PAGE(vm0_, pdpt, PDPT, CONFIG_SOS_ADDRESS_SPACE);
+DEFINE_PGTABLE_PAGE(vm0_, pd, PD, CONFIG_SOS_ADDRESS_SPACE);
+DEFINE_PGTABLE_PAGE(vm0_, pt, PT, CONFIG_SOS_ADDRESS_SPACE);
+
+static struct page uos_nworld_pml4_pages[CONFIG_MAX_VM_NUM - 1U][PML4_PAGE_NUM(CONFIG_UOS_ADDRESS_SPACE)];
+static struct page uos_nworld_pdpt_pages[CONFIG_MAX_VM_NUM - 1U][PDPT_PAGE_NUM(CONFIG_UOS_ADDRESS_SPACE)];
+static struct page uos_nworld_pd_pages[CONFIG_MAX_VM_NUM - 1U][PD_PAGE_NUM(CONFIG_UOS_ADDRESS_SPACE)];
+static struct page uos_nworld_pt_pages[CONFIG_MAX_VM_NUM - 1U][PT_PAGE_NUM(CONFIG_UOS_ADDRESS_SPACE)];
+static struct page uos_sworld_pml4_pages[CONFIG_MAX_VM_NUM - 1U][1UL];
+static struct page uos_sworld_pdpt_pages[CONFIG_MAX_VM_NUM - 1U][1UL];
+static struct page uos_sworld_pd_pages[CONFIG_MAX_VM_NUM - 1U][PD_PAGE_NUM(CONFIG_TRUSTY_RAM_SIZE)];
+static struct page uos_sworld_pt_pages[CONFIG_MAX_VM_NUM - 1U][PT_PAGE_NUM(CONFIG_TRUSTY_RAM_SIZE)];
```

Comment:
why not merge vm0_***_pages to uos_nworld_***_pages[] array? there is no difference in terms of EPT, And there is no special vm0 in partition mode.
in struct pgtable_pages_info { 
remove 
    struct page *nworld_pml4_base;
    struct page *nworld_pdpt_base;
    struct page *nworld_pd_base;
    struct page *nworld_pt_base;

add
int has_security_world:
to indicate if 
   struct page *sworld_pml4_base;
    struct page *sworld_pdpt_base;
    struct page *sworld_pd_base;
    struct page *sworld_pt_base;

are valid for this VM.

Reply:
Will update a patch according to the comments and send it to email list